dali/pipeline/data/dltensor.cc

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/pipeline/data/dltensor.h"
+
+namespace dali {
+
+DLDataType GetDLType(const TypeInfo &type) {
+  DLDataType dl_type{};
+  DALI_TYPE_SWITCH(type.id(), T,
+      dl_type.bits = sizeof(T) * 8;
+      dl_type.lanes = 1;
+      if (std::is_floating_point<T>::value) {
+        dl_type.code = kDLFloat;
+      } else if (std::is_unsigned<T>::value) {
+        dl_type.code = kDLUInt;
+      } else if (std::is_integral<T>::value) {
+        dl_type.code = kDLInt;
+      } else {
+        DALI_FAIL("This data type (" + type.name() + ") cannot be handled by DLTensor.");
+      })
+  return dl_type;
+}
+
+struct DLTensorResource {
+  kernels::TensorShape<> shape_;
+};
+
+void DLManagedTensorDeleter(DLManagedTensor *self) {
+  delete static_cast<DLTensorResource*>(self->manager_ctx);
+}
+
+void DLMTensorPtrDeleter(DLManagedTensor* dlm_tensor_ptr) {
+  if (dlm_tensor_ptr) {
+    dlm_tensor_ptr->deleter(dlm_tensor_ptr);
+    delete dlm_tensor_ptr;
+  }
+}
+
+DLMTensorPtr MakeDLTensor(void* data, const TypeInfo& type,
```

Comment:
Can we pass the custom deleter as an argument here, so in the future we can do some more elaborate things depending on the use case. Also it will provide the ability to test if the deleter works at all.

Reply:
Then when we expose this through the Python API we just need to prepare right deleter, the rest will be left untouched.

Reply:
I moved the DLTensorResource struct to the header and gave it a virtual destructor. Also, now a resource is passed as argument to the `MakeDLTensor`. I renamed `GetDLTensor` and `GetDLTensors` to `GetDLTesnorView` and `GetDLTensorListView` to mark that those return non-owning DLTensors. 
Now, when we want to make DLTensors with some kind of ownership we can reuse `MakeDLTensor` function and provide custom resource by defining a class deriving from DLTensorResource. I added an example to the tests.